design/0005-circuit-breaking.md

```diff
@@ -0,0 +1,150 @@
+# Title
+
+- Contribution Name: circuit-breaking
+- Implementation Owner: @Ashish-Bansal, Linkerd Team
+- Start Date: 2020-06-14
+- Target Date: N.A.
+- Linkerd Issue:
+  [linkerd/linkerd2#2846](https://github.com/linkerd/linkerd2/issues/2846)
+- Reviewers: @adleong
+
+## Summary
+
+[summary]: #summary
+
+We would like to implement circuit breaking functionality which would allow end
+users to configure mesh such that they can limit the impact of failures from
+faulty service over others and prevent cascading issues. By configuring circuit
+breaking, they will be able to remove failing service instances from load
+balancing and survive through it in a more graceful manner.
+
+## Problem Statement
+
+In a service mesh, services often make calls to other services. In case upstream
+service is unable to respond due to any reason, it's possible that these
+failures can cascade to other services in the cluster.
+
+If we are aware that specific service instance is not responding, we can call it
+unhealthy and it's better if we just stop sending "traffic"[0] to that instance
+rather than getting just another expected failure.
+
+Users should also be able to set max-ejection-percentage which provides max
+number of service instances which can be removed from load balancing for
+specific service.
+
+Along with that, if we detect that "almost all" the instances of service are not
+responding, then users should be able to cut off traffic to whole service as
+well. To quantize "almost all" term, user should be able to configure how many
+percentage of service instances would be removed from the load balancing list
+before whole traffic to that service would be circuit-breaked. This is in
+constrast with max-ejection-percentage and would be mutually exclusive.
+
+User should also be able to select ejection policy based on which any instance
+would be ejected from load balancing.
+
+It would also be somewhat helpful in case of canary deployments, when there's
+some regression and new build is raising failures.
+
+[0] The term 'traffic' means any kind of communication happening between two
+services. It can be HTTP requests or gRPC messages over HTTP or L4 TCP
+connections. But the circuit breaking functionality itself won't be protocol
+agnostic. Its implementation would be protocol specific and it's upto community
+which protocol they want to add support for and how circuit breaking will make
+sense in that protocol.
+
+## Design proposal
+
+[design-proposal]: #design-proposal
+
+Right now if I just consider ServiceProfile spec, IMO it seems to represent more
+of profile-for-http-service rather than profile-for-generic-service. May be it
+makes sense given that it supports only HTTP and grpc over HTTP as of now. Given
+current architecture, I feel it's best to provide circuit breaking functionality
+through service profiles. That means it would be limited to above two kind of
+services as of now.
+
+For TCP connections, according to original github
```

Comment:
TCP feels a little weird for circuit breaking. Maybe we should keep it out of scope for now? That'll help focus this a little better. To your earlier point, it feels like circuit breaking will need to be sorted out on a per-protocol level.